src/dawn/Optimizer/PassFieldVersioning.cpp

```diff
@@ -89,71 +91,165 @@ static void reportRaceCondition(const Statement& statement,
+  iir::Stage::DoMethodSmartPtr_t domethod =
+      make_unique<iir::DoMethod>(interval, context->getMetaData());
+  domethod->clearChildren();
+
+  addAssignmentToDoMethod(domethod, assignmentID, assigneeID, context, readExtents);
+
+  // Add the single do method to the new Stage
+  assignmentStage->clearChildren();
+  assignmentStage->addDoMethod(domethod);
+  for(auto& doMethod : assignmentStage->getChildren()) {
+    doMethod->update(iir::NodeUpdateType::level);
+  }
+  assignmentStage->update(iir::NodeUpdateType::level);
+  return assignmentStage;
+}
 
-PassFieldVersioning::PassFieldVersioning() : Pass("PassFieldVersioning", true), numRenames_(0) {}
+PassFieldVersioning::PassFieldVersioning(FieldVersioningPassMode mode)
+    : Pass("PassFieldVersioning", true), mode_(mode), numRenames_(0) {}
 
 bool PassFieldVersioning::run(
     const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) {
-  OptimizerContext* context = stencilInstantiation->getOptimizerContext();
-  numRenames_ = 0;
-
-  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
-    iir::Stencil& stencil = *stencilPtr;
-
-    // Iterate multi-stages backwards
-    int stageIdx = stencil.getNumStages() - 1;
-    for(auto multiStageRit = stencil.childrenRBegin(), multiStageRend = stencil.childrenREnd();
-        multiStageRit != multiStageRend; ++multiStageRit) {
-      iir::MultiStage& multiStage = (**multiStageRit);
-      iir::LoopOrderKind loopOrder = multiStage.getLoopOrder();
-
-      std::shared_ptr<iir::DependencyGraphAccesses> newGraph, oldGraph;
-      newGraph =
-          std::make_shared<iir::DependencyGraphAccesses>(stencilInstantiation->getMetaData());
-
-      // Iterate stages bottom -> top
-      for(auto stageRit = multiStage.childrenRBegin(), stageRend = multiStage.childrenREnd();
-          stageRit != stageRend; ++stageRit) {
-        iir::Stage& stage = (**stageRit);
-        iir::DoMethod& doMethod = stage.getSingleDoMethod();
-
-        // Iterate statements bottom -> top
-        for(int stmtIndex = doMethod.getChildren().size() - 1; stmtIndex >= 0; --stmtIndex) {
-          oldGraph = newGraph->clone();
-
-          auto& stmtAccessesPair = doMethod.getChildren()[stmtIndex];
-          newGraph->insertStatementAccessesPair(stmtAccessesPair);
-
-          // Try to resolve race-conditions by using double buffering if necessary
-          auto rc = fixRaceCondition(stencilInstantiation, newGraph.get(), stencil, doMethod,
-                                     loopOrder, stageIdx, stmtIndex);
-
-          if(rc == RCKind::RK_Unresolvable) {
-            // Nothing we can do ... bail out
-            return false;
-          } else if(rc == RCKind::RK_Fixed) {
-            // We fixed a race condition (this means some fields have changed and our current graph
-            // is invalid)
-            newGraph = oldGraph;
+  if(mode_ == FieldVersioningPassMode::FM_CreateVersion) {
```

Comment:
As said, I would propose to move it into separate pass and as our convention document the algorithm (Input/Output and what transformation applies to IIR)